tools/bro-to-vast/bro-to-vast.cpp

```diff
@@ -0,0 +1,304 @@
+      broker::vector result;
+      result.reserve(xs.size());
+      std::transform(xs.begin(), xs.end(), std::back_inserter(result),
+                     [](const auto& x) { return to_broker(x); });
+      return result;
+    },
+    [](const vast::set& xs) -> broker::data {
+      broker::set result;
+      std::transform(xs.begin(), xs.end(), std::inserter(result, result.end()),
+                     [](const auto& x) { return to_broker(x); });
+      return result;
+    },
+    [](const vast::map& xs) -> broker::data {
+      broker::table result;
+      auto f = [](const auto& x) { return std::pair{to_broker(x.first),
+                                                    to_broker(x.second)}; };
+      std::transform(xs.begin(), xs.end(), std::inserter(result, result.end()),
+                     f);
+      return result;
+    }
+  ), data);
+}
+
+// Constructs a result event for Bro from Broker data.
+broker::bro::Event make_bro_event(std::string name, broker::data x) {
+  broker::vector args(2);
+  args[0] = std::move(name);
+  args[1] = std::move(x);
+  return {"VAST::result", std::move(args)};
+}
+
+// Constructs a result event for Bro from a VAST event.
+broker::bro::Event make_bro_event(const vast::event& x) {
+  return make_bro_event(x.type().name(), to_broker(x.data()));
+}
+
+// A VAST writer that publishes the event it gets to a Bro endpoint.
+class bro_writer {
+public:
+  bro_writer() = default;
+
+  bro_writer(broker::endpoint& endpoint, std::string query_id)
+    : endpoint_{&endpoint},
+      query_id_{std::move(query_id)} {
+    // nop
+  }
+
+  caf::expected<void> write(const vast::event& x) {
+    std::cerr << '.';
+    endpoint_->publish(data_topic, make_bro_event(x));
+    return caf::no_error;
+  }
+
+  caf::expected<void> flush() {
+    return caf::no_error;
+  }
+
+  auto name() const {
+    return "bro-writer";
+  }
+
+private:
+  broker::endpoint* endpoint_;
+  std::string query_id_;
+};
+
+// A custom command that allows us to re-use VAST command dispatching logic in
+// order to issue a query that writes into a sink with a custom format.
+class bro_command : public vast::system::writer_command_base {
+public:
+  bro_command(broker::endpoint& endpoint)
+    : writer_command_base{nullptr, "bro"},
+      endpoint_{endpoint} {
+    // nop
+  }
+
+  // Sets the query ID to the UUID provided by Bro.
+  void query_id(std::string id) {
+    query_id_ = std::move(id);
+  }
+
+  /// Retrieves the current sink actor, which terminates when the exporter
+  /// corresponding to the issued query terminates.
+  caf::actor sink() const {
+    return sink_;
+  }
+
+protected:
+  caf::expected<caf::actor> make_sink(caf::scoped_actor& self,
+                                      const caf::config_value_map& options,
+                                      argument_iterator begin,
+                                      argument_iterator end) override {
+    VAST_UNUSED(options, begin, end);
```

Comment:
I think you've missed a `VAST_TRACE(VAST_ARG(options), VAST_ARG("args", begin, end));`.

Reply:
What do you mean by "missed"? I'm not aware of tracing guidelines at the moment.